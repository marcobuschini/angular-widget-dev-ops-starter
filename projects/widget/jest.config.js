module.exports = {
  preset: 'jest-preset-angular',
  setupFilesAfterEnv: ['<rootDir>/setup-jest.ts'],
  testEnvironment: 'jsdom',
  verbose: true,
  collectCoverage: true,
  coveragePathIgnorePatterns: ['/node_modules/', '/test/assets/'],
  reporters: [
    'default',
    [
      'jest-junit',
      {
        outputDirectory: 'tests/results',
        outputName: 'report.xml',
      },
    ],
    [
      'jest-html-reporters',
      {
        pageTitle: 'Test Report',
        filename: 'results.html',
        publicPath: 'tests/results',
      },
    ],
  ],
  coverageReporters: ['lcovonly', 'html', 'text'],
  coverageDirectory: 'tests/coverage',
  roots: ['<rootDir>/src'],
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.(html|svg)$',
    },
  },
  transform: {
    //'^.+\\.tsx?$': 'jest-preset-angular',
    '^.+\\.(ts|tsx|js|jsx|html|svg)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!@ngrx|(?!deck.gl)|ng-dynamic)'],
  moduleFileExtensions: [
    'ts',
    'tsx',
    'js',
    'jsx',
    'json',
    'node',
    'html',
    'mjs',
    'svg',
  ],
}
