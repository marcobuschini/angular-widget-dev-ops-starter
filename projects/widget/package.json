{
  "name": "widget",
  "version": "0.0.1",
  "description": "This is work in progress and it is not supposed to be used in production, test, or development",
  "author": {
    "name": "Marco Buschini",
    "email": "marco.buschini@gmail.com",
    "url": "https://marcobuschini.github.io/"
  },
  "license": "MIT",
  "peerDependencies": {
    "@angular/animations": "^12.2.11",
    "@angular/cdk": "~12.2.11",
    "@angular/common": "^12.2.11",
    "@angular/compiler": "^12.2.11",
    "@angular/core": "^12.2.11",
    "@angular/flex-layout": "^12.0.0-beta.35",
    "@angular/forms": "^12.2.11",
    "@angular/material": "^12.2.11",
    "@angular/platform-browser": "^12.2.11",
    "@angular/platform-browser-dynamic": "^12.2.11",
    "@angular/router": "^12.2.11",
    "rxjs": "^6.5.5",
    "zone.js": "~0.10.3"
  },
  "devDependencies": {
    "@angular-builders/jest": "^12.1.2",
    "@angular-devkit/build-angular":"^12.2.11",
    "@angular-devkit/build-ng-packagr": "0.1002.0",
    "@angular/cli": "^12.2.11",
    "@angular/compiler-cli":  "^12.2.11",
    "@angular/language-service":  "^12.2.11",
    "@compodoc/compodoc": "^1.1.11",
    "@types/jest": "^27.0.2",
    "@types/node": "^12.12.37",
    "codelyzer": "^5.2.2",
    "commitizen": "^4.0.4",
    "coveralls": "^3.1.0",
    "cz-conventional-changelog": "^3.1.0",
    "husky": "^4.2.15",
    "jest": "^27.3.1",
    "jest-html-reporters": "^2.1.6",
    "jest-preset-angular": "^10.1.0",
    "ng-packagr": "^12.2.5",
    "ts-jest": "^27.0.7",
    "ts-node": "~7.0.0",
    "tslint": "~5.15.0",
    "typescript": "~4.2.4"
  }
}
