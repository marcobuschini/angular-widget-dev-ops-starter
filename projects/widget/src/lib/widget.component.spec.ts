import {
  ComponentFixture,
  TestBed,
} from '@angular/core/testing'

import { expect, jest } from '@jest/globals'

import { WidgetComponent } from './widget.component'
import { Vendor } from './vendor'
import { provideHttpClientTesting } from '@angular/common/http/testing'
import { of } from 'rxjs'
import { ParkingSlot } from './parkingslot'
import { MatCardModule } from '@angular/material/card'
import { MatDividerModule } from '@angular/material/divider'
import { MatListModule } from '@angular/material/list'
import { BrowserModule, By } from '@angular/platform-browser'
import { provideHttpClient } from '@angular/common/http'
import { WidgetModule } from './widget.module'
import { WidgetService } from './widget.service'
import { FormsModule, ReactiveFormsModule } from '@angular/forms'
import { CommonModule } from '@angular/common'
import { MatButtonModule } from '@angular/material/button'

describe('WidgetComponent', () => {
  let component: WidgetComponent
  let fixture: ComponentFixture<WidgetComponent>

  let vendorSpy: any
  let slotsSpy: any
  let addSpy: any

  const dummyVendor: Vendor = {
    name: 'Test Vendor',
    description: 'This is a test vendor',
    features: [
      { name: 'Feature 1' },
      { name: 'Feature 2' },
      { name: 'Feature 3' },
    ],
  }

  const dummySlots: ParkingSlot[] = [
    {
      id: 0,
      name: 'Slot 1',
      features: ['Feature 1', 'Feature 2'],
    },
  ]

  const setTimeoutPromise = (milliseconds: number): Promise<void> =>
    new Promise(resolve => {
      setTimeout(resolve, milliseconds)
    })

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        CommonModule,
        MatButtonModule,
        MatCardModule,
        MatDividerModule,
        MatListModule,
        FormsModule,
        ReactiveFormsModule,
        WidgetModule,
        WidgetComponent,
      ],
      providers: [
        provideHttpClient(),
        provideHttpClientTesting(),
        WidgetService,
      ]
    }).compileComponents()
    fixture = TestBed.createComponent(WidgetComponent)
    component = fixture.componentInstance
  })

  beforeEach(() => {
    component.vendor = dummyVendor
    vendorSpy = jest
      .spyOn(component.service, 'getVendorFeatures')
      .mockReturnValue(of(dummyVendor.features))
    slotsSpy = jest
      .spyOn(component.service, 'getParkingSlots')
      .mockReturnValue(of(dummySlots))
    addSpy = jest.spyOn(component.buying, 'emit')
    component.ngOnInit()

    fixture.detectChanges()
  })

  afterEach(() => {
    component.ngOnDestroy()
  })

  it(
    'should create component with hidden slots, then show, then hide back',
    async () => {
      expect(component).toBeTruthy()

      component.ngOnInit()

      await setTimeoutPromise(1000)

      expect(vendorSpy).toHaveBeenCalled()
      expect(slotsSpy).toHaveBeenCalled()

      expect(fixture).toMatchSnapshot()

      let button = fixture.debugElement.queryAll(By.css('button'))[1]
        .nativeElement as HTMLButtonElement
      button.click()
      fixture.detectChanges()
      await setTimeoutPromise(1000)

      expect(fixture).toMatchSnapshot()

      button = fixture.debugElement.queryAll(By.css('button'))[1]
        .nativeElement as HTMLButtonElement
      button.click()
      fixture.detectChanges()
      await setTimeoutPromise(1000)

      expect(fixture).toMatchSnapshot()
    }
  )

  it('should add to cart', () => {
    component.ngOnInit()

    const button = fixture.debugElement.queryAll(By.css('button'))[0]
      .nativeElement as HTMLButtonElement
    button.click()
    fixture.detectChanges()
    expect(addSpy).toHaveBeenCalled()
  })
})
