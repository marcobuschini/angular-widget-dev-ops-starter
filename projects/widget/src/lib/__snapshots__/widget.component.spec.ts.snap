// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 1`] = `
<mlb-parking-widget
  buying={[Function OutputEmitterRef]}
  features={[Function Array]}
  features$={[Function SafeSubscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function SafeSubscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  vendor={[Function Function]}
>
  <mat-card
    class="mat-mdc-card mdc-card"
  >
    <mat-card-title
      class="mat-mdc-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-mdc-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-mdc-card-content"
    >
      <mat-list
        aria-disabled="false"
        class="mat-mdc-list mat-mdc-list-base mdc-list"
        role="list"
      >
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 1
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 2
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 3
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div>
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
          >
             Add to cart 
          </button>
          <button
            mat-stroked-button=""
          >
            Expand
          </button>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 2`] = `
<mlb-parking-widget
  buying={[Function OutputEmitterRef]}
  features={[Function Array]}
  features$={[Function SafeSubscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function SafeSubscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  vendor={[Function Function]}
>
  <mat-card
    class="mat-mdc-card mdc-card"
  >
    <mat-card-title
      class="mat-mdc-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-mdc-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-mdc-card-content"
    >
      <mat-list
        aria-disabled="false"
        class="mat-mdc-list mat-mdc-list-base mdc-list"
        role="list"
      >
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 1
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 2
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 3
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div>
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
          >
             Add to cart 
          </button>
          <button
            mat-stroked-button=""
          >
             Collapse 
          </button>
          <mat-list
            aria-disabled="false"
            class="mat-mdc-list mat-mdc-list-base mdc-list"
            role="list"
          >
            <mat-list-item
              aria-disabled="false"
              class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
              role="listitem"
            >
              <span
                class="mdc-list-item__content"
              >
                <span
                  class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
                >
                  Feature 1 
                </span>
              </span>
              <div
                class="mat-focus-indicator"
              />
            </mat-list-item>
            <mat-list-item
              aria-disabled="false"
              class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
              role="listitem"
            >
              <span
                class="mdc-list-item__content"
              >
                <span
                  class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
                >
                  Feature 2 
                </span>
              </span>
              <div
                class="mat-focus-indicator"
              />
            </mat-list-item>
          </mat-list>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 3`] = `
<mlb-parking-widget
  buying={[Function OutputEmitterRef]}
  features={[Function Array]}
  features$={[Function SafeSubscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function SafeSubscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  vendor={[Function Function]}
>
  <mat-card
    class="mat-mdc-card mdc-card"
  >
    <mat-card-title
      class="mat-mdc-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-mdc-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-mdc-card-content"
    >
      <mat-list
        aria-disabled="false"
        class="mat-mdc-list mat-mdc-list-base mdc-list"
        role="list"
      >
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 1
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 2
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
        <mat-list-item
          aria-disabled="false"
          class="mat-mdc-list-item mdc-list-item mat-mdc-list-item-single-line mdc-list-item--with-one-line"
          role="listitem"
        >
          <span
            class="mdc-list-item__content"
          >
            <span
              class="mat-mdc-list-item-unscoped-content mdc-list-item__primary-text"
            >
              Feature 3
            </span>
          </span>
          <div
            class="mat-focus-indicator"
          />
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div>
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
          >
             Add to cart 
          </button>
          <button
            mat-stroked-button=""
          >
            Expand
          </button>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;
